### PR TITLE
Check non-blocking connect() status in select_s()

### PR DESCRIPTION
**Pull request by jwt27**
Alternative to #79.

Work in progress - not yet tested!

### CODE REVIEW

File: src/select.c [hunk 1]
```diff
@@ -441,30 +441,23 @@ static __inline int listen_queued (Socket *socket)
...
-  /* A normal blocking socket. 'so_error' is set in
-   * connect(), a "ICMP Unreachable" callback or when RST received
-   * in pctcp.c. Otherwise 'so_error' is zero.
-   */
-  if (!(socket->so_state & SS_NBIO))
-     return (socket->so_error);
+  if (socket->so_state & SS_NBIO)
+  {
+    if (socket->so_options & SO_ACCEPTCONN) /* non-blocking listen sock */
+       return (0);
```
Comment (jwt27) [inline src/select.c:454]: Should all errors be ignored on a listen socket?

File: src/socket.c [hunk 2]
```diff
@@ -2274,6 +2274,40 @@ int _UDP6_listen (Socket *socket, const void *host, WORD port)
...
+  if (socket->so_error != EALREADY)
+    return (socket->so_error);
+
+  if (socket->so_state & SS_ISCONNECTED)
+    return (0);
+
+  if (socket->so_state & SS_CONN_REFUSED)
+    return (ECONNREFUSED);
+
+  if (chk_timeout(socket->nb_timer))
```
Comment (jwt27) [inline src/socket.c:2293]: For some reason this timer expires immediately.

Comment (jwt27) [reply src/socket.c:2293]: Oh I see.  `socket->delay` is set to 0 in a non-blocking socket.  So `socket->nbtimer` should be set from the default `sock_delay`.

**Comment by gvanem** [comment]
I've just restarted the AppVeoyr job on this PR. My bad.

File: src/select.c (hunk 1 above)
Comment (gvanem) [reply src/select.c:454]: Not sure. In what scenario should a non-blocking listen socket return an error?

Comment (jwt27) [reply src/select.c:454]: Right, it doesn't matter, because there aren't any.  I guess you could call `shutdown()` on it but that doesn't make much sense.

**Comment by jwt27** [comment]
OK, this works for me, but I don't quite like the idea of using `so_error` to keep track of the connection state.  `SS_ISCONNECTING` seems a better choice, but that is being set/cleared in a few other places already.  Will see if I can find a solution.

**Comment by gvanem** [comment]
> OK, this works for me, 

You mean for the C example you posted for PR-79?

**Comment by jwt27** [comment]
> > OK, this works for me,
> 
> You mean for the C example you posted for PR-79?

Yes, and the asio http_client example too.

**Comment by jwt27** [comment]
I checked what Linux does.  It never sets `EINPROGRESS`/`EALREADY` in `SO_ERROR` (it is always 0 after `connect()`), so we probably shouldn't do that either.  I'm curious to know how *BSD handles it though.

Also I'm wondering about `tcp_sock_daemon()`.  It toggles the connection status flags, but why?  Isn't it only for cleanup?

**Comment by jwt27** [comment]
Now using `SS_ISCONNECTED`, I think this is a much better solution.

I also checked how Linux handles `send()`/`recv()` after a non-blocking connect.  It fails with `EAGAIN`, and you can repeatedly call it until the connection succeeds.  But if you clear the non-blocking flag after `connect()`, then `send()`/`recv()` will wait for the connect to finish.  I copied that behavior here.

The last commit isn't 100% related to this issue, but it produces more specific errnos if we get an ICMP reject or the connection is reset.  It also prevents `SO_ERROR` from being ignored.

**Comment by gvanem** [comment]
> . It toggles the connection status flags, but why? 

Again, a long time since I wrote that. Can't remember the rationale before running a test-program.
Perhaps `tests/pull-request-79.c`?

**Comment by gvanem** [comment]
>  I copied that behavior here.

Good job. I'll merge this soon.

But it would be nice if you/me could test this with some real-life programs.
Like Wget, curl or Lynx if you're up for it. I have private Watt-32 makefiles for all these. But for Watt-32 on Windows.
I'm in no position to run djgpp on a DOS/Win-XP PC or an emulator. How do you run djgpp programs?

**Comment by jwt27** [comment]
> > . It toggles the connection status flags, but why?
> 
> Again, a long time since I wrote that. Can't remember the rationale before running a test-program. Perhaps `tests/pull-request-79.c`?

I *think* you meant for that to handle the non-blocking connect.  Don't really see what it would do otherwise.  The 500ms daemon tick rate seems rather slow for that though, so I moved the logic to `_sock_pending_connect()`, which also makes sure `SO_ERROR` is set to something meaningful when `SS_ISCONNECTING` is cleared.

> > I copied that behavior here.
> 
> Good job. I'll merge this soon.
> 
> But it would be nice if you/me could test this with some real-life programs. Like Wget, curl or Lynx if you're up for it. I have private Watt-32 makefiles for all these. But for Watt-32 on Windows. I'm in no position to run djgpp on a DOS/Win-XP PC or an emulator. How do you run djgpp programs?

I cross-compile on Debian or Win7, and then wget the binaries from a DOS machine.  If you could send me those makefiles, I'll see if I can get some of those programs running then.  Thanks!

**Comment by jwt27** [comment]
I got Lynx compiled but I'm not able to run it.  Something about ncurses failing to initialize.

Wget (latest version) won't compile, throws a long list of errors.

**Comment by gvanem** [comment]
> I got Lynx compiled but I'm not able to run it. Something about ncurses failing to initialize.

I'm using S-Lang in my home-brewed [Makefile.Windows](https://github.com/gvanem/Watt-32/files/11965326/Makefile.Windows.txt). And Lynx/Watt-32 still works fine AFAICS.

And for curl, I use these [Makefiles](https://github.com/gvanem/Watt-32/files/11965465/curl-Watt-32.zip):
```
unzip curl-Watt-32.zip
cd lib
make -f Makefile.Windows USE_WATT32=1 
cd ..\src
make -f Makefile.Windows USE_WATT32=1 
```

But using it; `curl.exe -v --wdebug www.watt-32.net`, I get the error this PR tries to solve ...

As for Wget, I've added an old version of it [here](https://github.com/gvanem/Watt-32/commit/ffa6712e908f1c7182a667f4f51f4461cbb6a06d). For Windows, use:
```
cd bin\Wget.182
nmake -f makefile.vc6 clean all
wget_vc.exe --wdebug -d -r -np www.curl.se
```

But since OpenSSL is not built-in (it's possible), it's not very useful.

**Comment by gvanem** [comment]
*I wrote*> Good job. I'll merge this soon.

Merged and tested with Lynx, Curl and Wget 1.82. Seems to work fine still. libcurl works much better:
```
curl www.watt-32.net/index.html --remote-name
lynx -dump index.html
...
Watt-32 tcp/ip Homepage
```
Or from a [bin/wget.182/makefile.vc6](bin/wget.182/makefile.vc6) build:
```
wget_vc.exe -d www.watt-32.net
lynx -dump index.html
...
Watt-32 tcp/ip Homepage
```

**Comment by jwt27** [comment]
Thanks for merging!

> > I got Lynx compiled but I'm not able to run it. Something about ncurses failing to initialize.
> 
> I'm using S-Lang in my home-brewed [Makefile.Windows](https://github.com/gvanem/Watt-32/files/11965326/Makefile.Windows.txt). And Lynx/Watt-32 still works fine AFAICS.

Just tried S-Lang but the latest version didn't want to compile for djgpp.

> And for curl, I use these [Makefiles](https://github.com/gvanem/Watt-32/files/11965465/curl-Watt-32.zip):

What version do these apply to?  I did get the latest version to build (just with `configure`).  Seems to work but it's *extremely* slow (3KB/s).

> > Wget (latest version) won't compile, throws a long list of errors.
> 
> Presumably since the real `<winsock*.h>` messes things up. Using `CFLAGS=-I$(WATT_ROOT)/inc/w32-fakes ...` and adding `-I$(WATT_ROOT)/inc` ahead of Gnulib's headers could work.

Doesn't seem to be Watt-related, it's just not compatible with djgpp ("please port gnulib to your platform" etc).  I see there is an msdos/config.h file now but I'm not sure how to use it in a cross compile.

**Pull request merged**